File: packages/gbnf-experiment/data/20260909T150124Z_a7cd2ee9/ported_implementation/src/grammar-graph/grammar-graph-types.ts
import type { GraphPointer } from './graph-pointer';
import type { Pointers } from './pointers';
import type { RuleRef } from './rule-ref';
import { validateNonEmpty } from '../utils/validate-non-empty';

export interface PrintOpts {
  pointers?: Pointers;
  colorize: (value: string | number, color: string) => string;
  showPosition?: boolean;
}

export type Range = [number, number];

export const RuleType = {
  CHAR: 'char',
  CHAR_EXCLUDE: 'char_exclude',
  END: 'end',
  REF: 'ref',
} as const;
export type RuleType = (typeof RuleType)[keyof typeof RuleType];

export class RuleEnd {
  type: typeof RuleType.END = RuleType.END;
}

export class RuleChar {
  type: typeof RuleType.CHAR = RuleType.CHAR;
  value: (number | Range)[];

  constructor(value: (number | Range)[]) {
    this.value = [...validateNonEmpty(value)];
  }
}

export class RuleCharExclude {
  type: typeof RuleType.CHAR_EXCLUDE = RuleType.CHAR_EXCLUDE;
  value: (number | Range)[];

  constructor(value: (number | Range)[]) {
    this.value = [...validateNonEmpty(value)];
  }
}

export type UnresolvedRule = RuleChar | RuleCharExclude | RuleRef | RuleEnd;

// ValidInput can either be a string, or a number indicating a code point.
// It CANNOT be a number representing a number; a number intended as input (like "8")
// should be passed in as a string.
export type ValidInput = string | number | number[];

// RuleRefs should never be exposed to the end user.
export type ResolvedRule = RuleCharExclude | RuleChar | RuleEnd;

export type ResolvedGraphPointer = GraphPointer<ResolvedRule>;

File: proposed-projects/round-trip-experiment/reverse/20260909T020947Z_ffb57550/20260910T054605Z_51904c27/ported_implementation/gbnf/grammar_graph/grammar_graph_types.py
from typing import Callable, List, Union

from ..utils.validate_non_empty import validate_non_empty
from .rule_type import RuleType

Colorize = Callable[[Union[str, int]], str]

# A range is a two element list of code points, `[start, end]`.
Range = List[int]

# ValidInput can either be a string, or a number indicating a code point.
# It CANNOT be a number representing a number; a number intended as input (like "8")
# should be passed in as a string.
ValidInput = Union[str, int, List[int]]


class RuleChar:
    def __init__(self, value: List[Union[int, Range]]):
        self.type = RuleType.CHAR
        self.value = validate_non_empty(list(value))

    def __eq__(self, other: object) -> bool:
        if type(other) is not type(self):
            return NotImplemented
        return self.value == other.value

    def __repr__(self) -> str:
        return f"RuleChar({self.value!r})"


class RuleCharExclude:
    def __init__(self, value: List[Union[int, Range]]):
        self.type = RuleType.CHAR_EXCLUDE
        self.value = validate_non_empty(list(value))

    def __eq__(self, other: object) -> bool:
        if type(other) is not type(self):
            return NotImplemented
        return self.value == other.value

    def __repr__(self) -> str:
        return f"RuleCharExclude({self.value!r})"


class RuleEnd:
    def __init__(self) -> None:
        self.type = RuleType.END

    def __eq__(self, other: object) -> bool:
        if type(other) is not type(self):
            return NotImplemented
        return True

    def __repr__(self) -> str:
        return "RuleEnd()"


def rule_char(value: List[Union[int, Range]]) -> RuleChar:
    return RuleChar(value)


def rule_char_exclude(value: List[Union[int, Range]]) -> RuleCharExclude:
    return RuleCharExclude(value)


def rule_end() -> RuleEnd:
    return RuleEnd()
